File: athar/bottom/spatial.py
"""Placement and lightweight spatial feature extraction for Phase 1."""

from __future__ import annotations

import math
from collections import defaultdict, deque
from dataclasses import dataclass

from .edge_policy import DOMAIN_GEOMETRY, DOMAIN_PLACEMENT, EDGE_INCLUDE
from .types import ClassifiedEdge, ParseResult

_PLACEMENT_SCALE = 1_000_000


@dataclass(frozen=True)
class SpatialFeature:
    placement: tuple[int, ...] | None
    centroid: tuple[float, float, float] | None
    aabb: tuple[float, float, float, float, float, float] | None


def build_spatial_features(parse_result: ParseResult, edges: list[ClassifiedEdge]) -> dict[int, SpatialFeature]:
    """Extract placement, centroid, and AABB for entities."""
    entities = parse_result.entities
    placement_adj = _build_adjacency(edges, domain=DOMAIN_PLACEMENT)
    geometry_adj = _build_adjacency(edges, domain=DOMAIN_GEOMETRY)

    out: dict[int, SpatialFeature] = {}
    transform_cache: dict[int, list[list[float]]] = {}
    for step_id, entity in entities.items():
        if not (entity.is_product or entity.is_spatial):
            continue
        matrix = _placement_matrix_for_entity(step_id, parse_result, placement_adj, transform_cache)
        placement_q = _quantize_matrix(matrix) if matrix is not None else None
        points = _collect_geometry_points(step_id, parse_result, geometry_adj)
        world_points = _transform_points(points, matrix) if matrix is not None else points
        centroid = _compute_centroid(world_points)
        aabb = _compute_aabb(world_points)
        if centroid is None and matrix is not None:
            centroid = (matrix[0][3], matrix[1][3], matrix[2][3])
        if aabb is None and centroid is not None:
            cx, cy, cz = centroid
            aabb = (cx, cy, cz, cx, cy, cz)
        out[step_id] = SpatialFeature(placement=placement_q, centroid=centroid, aabb=aabb)
    return out


def _build_adjacency(edges: list[ClassifiedEdge], *, domain: str) -> dict[int, list[int]]:
    out: dict[int, list[int]] = defaultdict(list)
    for edge in edges:
        if edge.classification != EDGE_INCLUDE:
            continue
        if edge.domain != domain:
            continue
        out[edge.source_step].append(edge.target_step)
    for neighbors in out.values():
        neighbors.sort()
    return out


def _placement_matrix_for_entity(
    step_id: int,
    parse_result: ParseResult,
    placement_adj: dict[int, list[int]],
    cache: dict[int, list[list[float]]],
) -> list[list[float]] | None:
    entities = parse_result.entities
    entity = entities.get(step_id)
    if entity is None:
        return None

    for ref in entity.refs:
        if ref.attr_name == "ObjectPlacement":
            return _resolve_local_placement(ref.target_step, parse_result, placement_adj, cache, seen=set())
    return None


def _resolve_local_placement(
    placement_step: int,
    parse_result: ParseResult,
    placement_adj: dict[int, list[int]],
    cache: dict[int, list[list[float]]],
    *,
    seen: set[int],
) -> list[list[float]] | None:
    if placement_step in cache:
        return cache[placement_step]
    if placement_step in seen:
        return None
    seen.add(placement_step)

    entities = parse_result.entities
    placement = entities.get(placement_step)
    if placement is None:
        return None

    rel_to_step = _first_ref_target(placement, "PlacementRelTo")
    relative_step = _first_ref_target(placement, "RelativePlacement")

    base = _identity_matrix()
    if rel_to_step is not None:
        parent = _resolve_local_placement(rel_to_step, parse_result, placement_adj, cache, seen=seen)
        if parent is not None:
            base = parent

    local = _axis2placement_matrix(relative_step, parse_result) if relative_step is not None else _identity_matrix()
    matrix = _matmul(base, local)
    cache[placement_step] = matrix
    return matrix


def _axis2placement_matrix(step_id: int, parse_result: ParseResult) -> list[list[float]]:
    entities = parse_result.entities
    placement = entities.get(step_id)
    if placement is None:
        return _identity_matrix()

    location = _point_from_ref(_first_ref_target(placement, "Location"), parse_result)
    z_axis = _direction_from_ref(_first_ref_target(placement, "Axis"), parse_result, default=(0.0, 0.0, 1.0))
    x_axis = _direction_from_ref(_first_ref_target(placement, "RefDirection"), parse_result, default=(1.0, 0.0, 0.0))
    z = _normalize(z_axis)
    x = _normalize(_orthogonalize(x_axis, z))
    y = _normalize(_cross(z, x))
    x = _normalize(_cross(y, z))
    tx, ty, tz = location

    return [
        [x[0], y[0], z[0], tx],
        [x[1], y[1], z[1], ty],
        [x[2], y[2], z[2], tz],
        [0.0, 0.0, 0.0, 1.0],
    ]


def _collect_geometry_points(step_id: int, parse_result: ParseResult, geometry_adj: dict[int, list[int]]) -> list[tuple[float, float, float]]:
    entities = parse_result.entities
    queue = deque([step_id])
    seen: set[int] = set()
    points: list[tuple[float, float, float]] = []
    while queue:
        current = queue.popleft()
        if current in seen:
            continue
        seen.add(current)
        entity = entities.get(current)
        if entity is None:
            continue
        if entity.entity_type == "IfcCartesianPoint":
            parsed = _coords_from_point_entity(entity)
            if parsed is not None:
                points.append(parsed)
        for nxt in geometry_adj.get(current, []):
            if nxt not in seen:
                queue.append(nxt)
    return points


def _coords_from_point_entity(entity) -> tuple[float, float, float] | None:
    coords = entity.attributes.get("Coordinates")
    if not isinstance(coords, dict):
        return None
    items = coords.get("items")
    if not isinstance(items, list) or not items:
        return None
    raw = [_scalar_float(item) for item in items]
    if any(v is None for v in raw):
        return None
    xyz = [float(v) for v in raw if v is not None]
    while len(xyz) < 3:
        xyz.append(0.0)
    return (xyz[0], xyz[1], xyz[2])


def _point_from_ref(step_id: int | None, parse_result: ParseResult) -> tuple[float, float, float]:
    if step_id is None:
        return (0.0, 0.0, 0.0)
    entity = parse_result.entities.get(step_id)
    if entity is None:
        return (0.0, 0.0, 0.0)
    parsed = _coords_from_point_entity(entity)
    return parsed if parsed is not None else (0.0, 0.0, 0.0)


def _direction_from_ref(
    step_id: int | None,
    parse_result: ParseResult,
    *,
    default: tuple[float, float, float],
) -> tuple[float, float, float]:
    if step_id is None:
        return default
    entity = parse_result.entities.get(step_id)
    if entity is None:
        return default
    ratios = entity.attributes.get("DirectionRatios")
    if not isinstance(ratios, dict):
        return default
    items = ratios.get("items")
    if not isinstance(items, list) or not items:
        return default
    raw = [_scalar_float(item) for item in items]
    if any(v is None for v in raw):
        return default
    xyz = [float(v) for v in raw if v is not None]
    while len(xyz) < 3:
        xyz.append(0.0)
    return (xyz[0], xyz[1], xyz[2])


def _scalar_float(value) -> float | None:
    if not isinstance(value, dict):
        return None
    if value.get("kind") == "int":
        return float(value.get("value"))
    if value.get("kind") == "real_q":
        return float(value.get("value", 0)) / 1_000_000.0
    return None


def _compute_centroid(points: list[tuple[float, float, float]]) -> tuple[float, float, float] | None:
    if not points:
        return None
    sx = sum(p[0] for p in points)
    sy = sum(p[1] for p in points)
    sz = sum(p[2] for p in points)
    n = float(len(points))
    return (sx / n, sy / n, sz / n)


def _compute_aabb(points: list[tuple[float, float, float]]) -> tuple[float, float, float, float, float, float] | None:
    if not points:
        return None
    xs = [p[0] for p in points]
    ys = [p[1] for p in points]
    zs = [p[2] for p in points]
    return (min(xs), min(ys), min(zs), max(xs), max(ys), max(zs))


def _transform_points(
    points: list[tuple[float, float, float]],
    matrix: list[list[float]],
) -> list[tuple[float, float, float]]:
    return [_transform_point(matrix, point) for point in points]


def _transform_point(matrix: list[list[float]], point: tuple[float, float, float]) -> tuple[float, float, float]:
    x, y, z = point
    tx = matrix[0][0] * x + matrix[0][1] * y + matrix[0][2] * z + matrix[0][3]
    ty = matrix[1][0] * x + matrix[1][1] * y + matrix[1][2] * z + matrix[1][3]
    tz = matrix[2][0] * x + matrix[2][1] * y + matrix[2][2] * z + matrix[2][3]
    tw = matrix[3][0] * x + matrix[3][1] * y + matrix[3][2] * z + matrix[3][3]
    if abs(tw) > 1e-12 and abs(tw - 1.0) > 1e-12:
        return (tx / tw, ty / tw, tz / tw)
    return (tx, ty, tz)


def _quantize_matrix(matrix: list[list[float]]) -> tuple[int, ...]:
    q: list[int] = []
    for row in matrix:
        for value in row:
            q.append(int(round(value * _PLACEMENT_SCALE)))
    return tuple(q)


def _first_ref_target(entity, attr_name: str) -> int | None:
    for ref in entity.refs:
        if ref.attr_name == attr_name:
            return ref.target_step
    return None


def _identity_matrix() -> list[list[float]]:
    return [
        [1.0, 0.0, 0.0, 0.0],
        [0.0, 1.0, 0.0, 0.0],
        [0.0, 0.0, 1.0, 0.0],
        [0.0, 0.0, 0.0, 1.0],
    ]


def _matmul(a: list[list[float]], b: list[list[float]]) -> list[list[float]]:
    out = [[0.0] * 4 for _ in range(4)]
    for i in range(4):
        for j in range(4):
            out[i][j] = sum(a[i][k] * b[k][j] for k in range(4))
    return out


def _normalize(v: tuple[float, float, float]) -> tuple[float, float, float]:
    norm = math.sqrt(v[0] ** 2 + v[1] ** 2 + v[2] ** 2)
    if norm <= 1e-12:
        return (0.0, 0.0, 0.0)
    return (v[0] / norm, v[1] / norm, v[2] / norm)


def _cross(a: tuple[float, float, float], b: tuple[float, float, float]) -> tuple[float, float, float]:
    return (
        a[1] * b[2] - a[2] * b[1],
        a[2] * b[0] - a[0] * b[2],
        a[0] * b[1] - a[1] * b[0],
    )


def _orthogonalize(v: tuple[float, float, float], axis: tuple[float, float, float]) -> tuple[float, float, float]:
    dot = v[0] * axis[0] + v[1] * axis[1] + v[2] * axis[2]
    return (
        v[0] - dot * axis[0],
        v[1] - dot * axis[1],
        v[2] - dot * axis[2],
    )

File: athar/_native/src/spatial.rs
//! Native port of `athar/bottom/spatial.py`: resolve `ObjectPlacement` chains
//! into quantized world matrices, and compute centroid + AABB from the
//! geometry-domain point cloud.

use std::collections::{HashMap, HashSet, VecDeque};

use crate::canon::Entity;

const PLACEMENT_SCALE: f64 = 1_000_000.0;

pub struct SpatialFeature {
    pub placement: Option<Vec<i64>>,
    pub centroid: Option<[f64; 3]>,
    pub aabb: Option<[f64; 6]>,
}

type Matrix = [[f64; 4]; 4];

fn identity() -> Matrix {
    [
        [1.0, 0.0, 0.0, 0.0],
        [0.0, 1.0, 0.0, 0.0],
        [0.0, 0.0, 1.0, 0.0],
        [0.0, 0.0, 0.0, 1.0],
    ]
}

pub struct SpatialCtx<'a> {
    pub by_id: &'a HashMap<i64, &'a Entity>,
    pub geometry_adj: &'a HashMap<i64, Vec<i64>>,
    pub point_coords: &'a HashMap<i64, [f64; 3]>,
    pub dir_ratios: &'a HashMap<i64, [f64; 3]>,
}

pub fn build_spatial_features(
    entities: &[Entity],
    ctx: &SpatialCtx,
) -> HashMap<i64, SpatialFeature> {
    let mut out = HashMap::new();
    let mut transform_cache: HashMap<i64, Matrix> = HashMap::new();

    for ent in entities {
        if !(ent.is_product || ent.is_spatial) {
            continue;
        }
        let matrix = placement_matrix_for_entity(ent, ctx, &mut transform_cache);
        let placement_q = matrix.map(quantize_matrix);

        let points = collect_geometry_points(ent.step_id, ctx);
        let world_points: Vec<[f64; 3]> = match &matrix {
            Some(m) => points.iter().map(|p| transform_point(m, *p)).collect(),
            None => points,
        };
        let mut centroid = compute_centroid(&world_points);
        let mut aabb = compute_aabb(&world_points);
        if centroid.is_none() {
            if let Some(m) = &matrix {
                centroid = Some([m[0][3], m[1][3], m[2][3]]);
            }
        }
        if aabb.is_none() {
            if let Some(c) = centroid {
                aabb = Some([c[0], c[1], c[2], c[0], c[1], c[2]]);
            }
        }
        out.insert(
            ent.step_id,
            SpatialFeature {
                placement: placement_q,
                centroid,
                aabb,
            },
        );
    }
    out
}

fn first_ref_target(ent: &Entity, attr: &str) -> Option<i64> {
    ent.refs.iter().find(|r| &*r.attr_name == attr).map(|r| r.target)
}

fn placement_matrix_for_entity(
    ent: &Entity,
    ctx: &SpatialCtx,
    cache: &mut HashMap<i64, Matrix>,
) -> Option<Matrix> {
    let placement_step = first_ref_target(ent, "ObjectPlacement")?;
    let mut seen = HashSet::new();
    resolve_local_placement(placement_step, ctx, cache, &mut seen)
}

fn resolve_local_placement(
    placement_step: i64,
    ctx: &SpatialCtx,
    cache: &mut HashMap<i64, Matrix>,
    seen: &mut HashSet<i64>,
) -> Option<Matrix> {
    if let Some(m) = cache.get(&placement_step) {
        return Some(*m);
    }
    if seen.contains(&placement_step) {
        return None;
    }
    seen.insert(placement_step);

    let placement = ctx.by_id.get(&placement_step)?;
    let rel_to = first_ref_target(placement, "PlacementRelTo");
    let relative = first_ref_target(placement, "RelativePlacement");

    let mut base = identity();
    if let Some(parent_step) = rel_to {
        if let Some(parent) = resolve_local_placement(parent_step, ctx, cache, seen) {
            base = parent;
        }
    }
    let local = match relative {
        Some(step) => axis2placement_matrix(step, ctx),
        None => identity(),
    };
    let matrix = matmul(&base, &local);
    cache.insert(placement_step, matrix);
    Some(matrix)
}

fn axis2placement_matrix(step: i64, ctx: &SpatialCtx) -> Matrix {
    let placement = match ctx.by_id.get(&step) {
        Some(p) => p,
        None => return identity(),
    };
    let location = first_ref_target(placement, "Location")
        .and_then(|id| ctx.point_coords.get(&id).copied())
        .unwrap_or([0.0, 0.0, 0.0]);
    let z_axis = first_ref_target(placement, "Axis")
        .and_then(|id| ctx.dir_ratios.get(&id).copied())
        .unwrap_or([0.0, 0.0, 1.0]);
    let x_axis = first_ref_target(placement, "RefDirection")
        .and_then(|id| ctx.dir_ratios.get(&id).copied())
        .unwrap_or([1.0, 0.0, 0.0]);

    let z = normalize(z_axis);
    let mut x = normalize(orthogonalize(x_axis, z));
    let y = normalize(cross(z, x));
    x = normalize(cross(y, z));
    let [tx, ty, tz] = location;

    [
        [x[0], y[0], z[0], tx],
        [x[1], y[1], z[1], ty],
        [x[2], y[2], z[2], tz],
        [0.0, 0.0, 0.0, 1.0],
    ]
}

fn collect_geometry_points(step_id: i64, ctx: &SpatialCtx) -> Vec<[f64; 3]> {
    let mut queue = VecDeque::new();
    queue.push_back(step_id);
    let mut seen = HashSet::new();
    let mut points = Vec::new();
    while let Some(current) = queue.pop_front() {
        if !seen.insert(current) {
            continue;
        }
        if let Some(ent) = ctx.by_id.get(&current) {
            if ent.keyword == "IFCCARTESIANPOINT" {
                if let Some(c) = ctx.point_coords.get(&current) {
                    points.push(*c);
                }
            }
        }
        if let Some(neighbors) = ctx.geometry_adj.get(&current) {
            for &nxt in neighbors {
                if !seen.contains(&nxt) {
                    queue.push_back(nxt);
                }
            }
        }
    }
    points
}

fn transform_point(m: &Matrix, p: [f64; 3]) -> [f64; 3] {
    let [x, y, z] = p;
    let tx = m[0][0] * x + m[0][1] * y + m[0][2] * z + m[0][3];
    let ty = m[1][0] * x + m[1][1] * y + m[1][2] * z + m[1][3];
    let tz = m[2][0] * x + m[2][1] * y + m[2][2] * z + m[2][3];
    let tw = m[3][0] * x + m[3][1] * y + m[3][2] * z + m[3][3];
    if tw.abs() > 1e-12 && (tw - 1.0).abs() > 1e-12 {
        [tx / tw, ty / tw, tz / tw]
    } else {
        [tx, ty, tz]
    }
}

fn compute_centroid(points: &[[f64; 3]]) -> Option<[f64; 3]> {
    if points.is_empty() {
        return None;
    }
    let n = points.len() as f64;
    let mut s = [0.0f64; 3];
    for p in points {
        s[0] += p[0];
        s[1] += p[1];
        s[2] += p[2];
    }
    Some([s[0] / n, s[1] / n, s[2] / n])
}

fn compute_aabb(points: &[[f64; 3]]) -> Option<[f64; 6]> {
    if points.is_empty() {
        return None;
    }
    let mut min = points[0];
    let mut max = points[0];
    for p in &points[1..] {
        for i in 0..3 {
            if p[i] < min[i] {
                min[i] = p[i];
            }
            if p[i] > max[i] {
                max[i] = p[i];
            }
        }
    }
    Some([min[0], min[1], min[2], max[0], max[1], max[2]])
}

fn quantize_matrix(m: Matrix) -> Vec<i64> {
    let mut q = Vec::with_capacity(16);
    for row in m.iter() {
        for &v in row.iter() {
            q.push((v * PLACEMENT_SCALE).round() as i64);
        }
    }
    q
}

fn matmul(a: &Matrix, b: &Matrix) -> Matrix {
    let mut out = [[0.0f64; 4]; 4];
    for i in 0..4 {
        for j in 0..4 {
            let mut s = 0.0;
            for k in 0..4 {
                s += a[i][k] * b[k][j];
            }
            out[i][j] = s;
        }
    }
    out
}

fn normalize(v: [f64; 3]) -> [f64; 3] {
    let norm = (v[0] * v[0] + v[1] * v[1] + v[2] * v[2]).sqrt();
    if norm <= 1e-12 {
        [0.0, 0.0, 0.0]
    } else {
        [v[0] / norm, v[1] / norm, v[2] / norm]
    }
}

fn cross(a: [f64; 3], b: [f64; 3]) -> [f64; 3] {
    [
        a[1] * b[2] - a[2] * b[1],
        a[2] * b[0] - a[0] * b[2],
        a[0] * b[1] - a[1] * b[0],
    ]
}

fn orthogonalize(v: [f64; 3], axis: [f64; 3]) -> [f64; 3] {
    let dot = v[0] * axis[0] + v[1] * axis[1] + v[2] * axis[2];
    [
        v[0] - dot * axis[0],
        v[1] - dot * axis[1],
        v[2] - dot * axis[2],
    ]
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::canon::{Entity, RefOut};

    fn ent(step_id: i64, keyword: &str, is_product: bool, refs: &[(&str, i64)]) -> Entity {
        Entity {
            step_id,
            keyword: keyword.to_string(),
            canonical_class: keyword.to_string(),
            is_product,
            is_spatial: false,
            guid: None,
            name: None,
            geom_parts: Vec::new(),
            data_parts: Vec::new(),
            refs: refs
                .iter()
                .map(|(a, t)| RefOut { attr_name: std::rc::Rc::from(*a), target: *t })
                .collect(),
            data_facts: Vec::new(),
        }
    }

    #[test]
    fn placement_chain_yields_world_space_geometry() {
        // Wall placed at (10,0,0); its geometry point is local (1,2,3). The
        // centroid/AABB must be in world space (11,2,3) and the placement's
        // X translation must quantize to 10 * 1e6.
        let entities = vec![
            ent(1, "IFCWALL", true, &[("ObjectPlacement", 2)]),
            ent(2, "IFCLOCALPLACEMENT", false, &[("RelativePlacement", 3)]),
            ent(3, "IFCAXIS2PLACEMENT3D", false, &[("Location", 4)]),
            ent(4, "IFCCARTESIANPOINT", false, &[]),
            ent(5, "IFCCARTESIANPOINT", false, &[]),
        ];
        let by_id: HashMap<i64, &Entity> = entities.iter().map(|e| (e.step_id, e)).collect();
        let mut geometry_adj: HashMap<i64, Vec<i64>> = HashMap::new();
        geometry_adj.insert(1, vec![5]);
        let mut point_coords: HashMap<i64, [f64; 3]> = HashMap::new();
        point_coords.insert(4, [10.0, 0.0, 0.0]);
        point_coords.insert(5, [1.0, 2.0, 3.0]);
        let dir_ratios: HashMap<i64, [f64; 3]> = HashMap::new();

        let ctx = SpatialCtx {
            by_id: &by_id,
            geometry_adj: &geometry_adj,
            point_coords: &point_coords,
            dir_ratios: &dir_ratios,
        };
        let feats = build_spatial_features(&entities, &ctx);
        let f = feats.get(&1).expect("wall feature");

        assert_eq!(f.centroid, Some([11.0, 2.0, 3.0]));
        assert_eq!(f.aabb, Some([11.0, 2.0, 3.0, 11.0, 2.0, 3.0]));
        let placement = f.placement.as_ref().expect("placement matrix");
        assert_eq!(placement[3], 10_000_000); // row 0, col 3 = X translation * 1e6
    }
}
